### ISSUE TITLE
[question] Why does test_requires is installed when not testing?

### ISSUE THREAD

**Issue by uilianries** [conan-io/conan#11905]
The enigmatic feature `test_requires` is mentioned in Docs in some points:

https://docs.conan.io/en/latest/reference/commands/creator/create.html?highlight=test_requires
https://docs.conan.io/en/latest/reference/conanfile/dependencies.html?highlight=test_requires#iterating-dependencies
https://docs.conan.io/en/latest/migrating_to_2.0/recipes.html?highlight=test_requires#requirements
https://docs.conan.io/en/latest/devtools/build_requires.html?highlight=test_requires#build-and-host-contexts

There is no specif section about that feature, but is clear that's a shortcut to `self.build_requires(<ref>, force_host_context=True)`.

However, as it should be combined when testing a package with unit tests or similar, it would be nice installing test requirements only when indeed running tests. For instance:

```python
def build_requirements(self):
    self.test_requires("gtest/1.12.1")

def build(self):
    cmake = CMake(self)
    cmake.configure()
    cmake.build()
    cmake.test()
```

`conan create . foobar/0.1.0 -pr:h=default -pr:b=default -c tools.build:skip_test=True`

This command tells to Conan to skip the test step, which means, we don't need to install GTest neither run our unit tests. However, Conan always install gtest as it's a regular build requirement.

So, you can tell, but, it built all tests, only didn't execute them ... But again, we can avoid building them by:

```python
def generate(self):
    tc = CMakeToolchain(self)
    # Maybe Conan manages it automatically but I didn't check
    tc.variables["BUILD_TESTING"] = self.conf_info.get("tools.build:skip_test", check_type=bool)
    tc.generate()
```

Of course, users can apply the same rule to avoid installing gtest, but it seems be a dubious as we only want to install a requirement for unit tests.  

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

**Comment by lasote** [comment]
About skipping the `test_requires()` based on `tools.build:skip_test` I don't think that is a good idea to do it, sounds a bit tricky and hidden.

About the toolchain setting the variable `BUILD_TESTING` might be indeed a good idea 👍🏼

**Comment by memsharded** [comment]
> About skipping the test_requires() based on tools.build:skip_test I don't think that is a good idea to do it, sounds a bit tricky and hidden.

Agree, Conan 2.0 constructs the whole graph always, not skipping even tool_requires. It will only skip fetching binaries for packages when they are not needed (BINARY_SKIP). Maybe we want to consider this in the future, but at this step sounds premature optimization, not building and running tests is the way to go first, so the BUILD_TESTING sounds right too.

**Comment by franramirez688** [comment]
Closed by https://github.com/conan-io/conan/pull/11935.

It will be available in the next Conan 1.52.
